pending.c

```diff
@@ -98,14 +144,132 @@ void w_pending_coll_unlock(struct watchman_pending_collection *coll) {
+// and it is non-trivial to add that functionality right now.
+// When we lop-off a portion of a tree that we're going to analyze
+// recursively, we have to iterate each leaf and explicitly delete
+// that leaf.
+// Since deletion invalidates the iteration state we have to signal
+// to stop iteration after each deletion and then retry the prefix
+// deletion.
+// We use this kid_context state to pass down the required information
+// to the iterator callback so that we adjust the overall state correctly.
+struct kid_context {
+  w_string_t *root;
+  struct watchman_pending_collection *coll;
+};
+
+// We need to guarantee that we never collapse a cookie notification
+// out of the pending list, because we absolutely must observe it coming
+// in via the kernel notification mechanism in order for synchronization
+// to be correct.
+// Since we don't have a w_root_t available, we can't tell what the
+// precise cookie prefix is for the current pending list here, so
+// we do a substring match.  Not the most elegant thing in the world.
+static inline bool is_possibly_a_cookie(w_string_t *path) {
+  return w_string_contains_cstr_len(path, WATCHMAN_COOKIE_PREFIX,
+                                    sizeof(WATCHMAN_COOKIE_PREFIX) - 1);
+}
+
+// This is the iterator callback we use to prune out obsoleted leaves.
+// We need to compare the prefix to make sure that we don't delete
+// a sibling node by mistake (see commentary on the is_path_prefix
+// function for more on that).
+static int delete_kids(void *data, const unsigned char *key, uint32_t key_len,
+                       void *value) {
+  struct kid_context *ctx = data;
+  struct watchman_pending_fs *p = value;
+  unused_parameter(value);
+
+  if ((p->flags & W_PENDING_CRAWL_ONLY) == 0 && key_len > ctx->root->len &&
+      is_path_prefix((const char *)key, key_len, ctx->root->len) &&
+      !is_possibly_a_cookie(p->path)) {
+    // Unlink the child from the pending index.
+    unlink_item(ctx->coll, p);
+    // and completely free it.
+
+#if 0
+    {
+      char flags_label[128];
+      w_expand_flags(kflags, p->flags, flags_label, sizeof(flags_label));
+      w_log(W_LOG_ERR,
+            "delete_kids: remove %s %.*s because %.*s obsoleted it\n",
+            flags_label, p->path->len, p->path->buf, ctx->root->len,
+            ctx->root->buf);
+    }
+#endif
+
+    w_pending_fs_free(p);
+
+    // Remove it from the art tree also.
+    art_delete(&ctx->coll->tree, key, key_len);
+
+    // Stop iteration because we just invalidated the iterator state
+    // by modifying the tree mid-iteration.
+    return 1;
+  }
+
+  return 0;
+}
+
+// if there are any entries that are obsoleted by a recursive insert,
+// walk over them now and mark them as ignored.
+static void prune_obsoleted_children(struct watchman_pending_collection *coll,
+                                     w_string_t *path, int flags) {
+  if ((flags & (W_PENDING_RECURSIVE | W_PENDING_CRAWL_ONLY)) ==
```

Comment:
As a casual reader, I'd be surprised that there's a condition in here -- it might make more sense to call this `maybe_prune_obsoleted_children`